src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -297,8 +318,96 @@ internal static bool TryGetMinimalCSharpRange(RazorCodeDocument codeDocument, Ra
+
+        var count = responseData.Sum(r => r.Length);
+        var data = new int[count];
+        var dataIndex = 0;
+        var lastTokenLine = 0;
+
+        for (var i = 0; i < responseData.Length; i++)
+        {
+            var curData = responseData[i];
+
+            if (curData.Length == 0)
+            {
+                continue;
+            }
+
+            Array.Copy(curData, 0, data, dataIndex, curData.Length);
+            if (i != 0)
+            {
+                // The first two items in result.Data will potentially need it's line/col offset modified
+                var lineDelta = data[dataIndex] - lastTokenLine;
+
+                // Update the first line copied over from curData
+                data[dataIndex] = lineDelta;
+
+                // Update the first column copied over from curData if on the same line as the previous token
+                if (lineDelta == 0)
+                {
+                    var lastTokenCol = 0;
+
+                    // Walk back accumulating column deltas until we find a start column (indicated by it's line offset being non-zero)
+                    for (var j = dataIndex - RazorSemanticTokensInfoService.TokenSize; j >= 0; j -= RazorSemanticTokensInfoService.TokenSize)
+                    {
+                        lastTokenCol += data[dataIndex + 1];
+                        if (data[dataIndex] != 0)
+                        {
+                            break;
+                        }
+                    }
+
+                    data[dataIndex + 1] -= lastTokenCol;
+                }
+            }
+
+            lastTokenLine = 0;
+            for (var j = 0; j < curData.Length; j += RazorSemanticTokensInfoService.TokenSize)
+            {
+                lastTokenLine += curData[j];
+            }
+
+            dataIndex += curData.Length;
+        }
+
+        return data;
+    }
+
+    // Internal for testing only
+    internal static bool TryGetCSharpRanges(RazorCodeDocument codeDocument, Range razorRange, [NotNullWhen(true)] out Range[]? ranges)
+    {
+        using var _ = ListPool<Range>.GetPooledObject(out var csharpRanges);
+        var csharpSourceText = codeDocument.GetCSharpSourceText();
+        var sourceText = codeDocument.GetSourceText();
+        var textSpan = razorRange.AsTextSpan(sourceText);
+        var csharpDoc = codeDocument.GetCSharpDocument();
+        var foundRanges = false;
+
+        // We want to find the min and max C# source mapping that corresponds with our Razor range.
+        foreach (var mapping in csharpDoc.SourceMappings)
+        {
+            var mappedTextSpan = mapping.OriginalSpan.AsTextSpan();
+
+            if (textSpan.OverlapsWith(mappedTextSpan))
+            {
+                foundRanges = true;
+                var mappedRange = mapping.GeneratedSpan.AsRange(csharpSourceText);
+                csharpRanges.Add(mappedRange);
+            }
+        }
+
+        ranges = foundRanges ? csharpRanges.ToArray() : null;
```

Comment:
>foundRanges [](http://example.com/codeflow?start=17&length=11)

nit: consider getting rid of foundRanges and just doing something like:

ranges = csharpRanges.Count > 0 ? csharpRanges.ToArray() : null;